ntrans_gui.py

```diff
@@ -0,0 +1,268 @@
+            font=("TkDefaultFont", 16),
+            foreground="yellow",
+            background="black",
+            padding=30,
+        )
+        generate_dictionary = ttk.Button(
+            mainframe,
+            command=self.generate_ntrans_dictionary,
+            text="Generate N-Trans Dictionary",
+            style="W.TButton",
+        )
+
+        # About / Help
+        about_help_buttonframe = ttk.Frame(mainframe)
+
+        about_button = ttk.Button(
+            about_help_buttonframe,
+            command=self.open_about_window,
+            text="About",
+            style="",
+        )  # TODO: Make command into lambda instantiating about-class
+        help_button = ttk.Button(
+            about_help_buttonframe, command=self.open_help_page, text="Help"
+        )
+
+        # Black turn off formatting
+        # fmt: off
+
+        # ==================
+        # Widget placement
+        # ==================
+        header.grid(column=0, row=0, columnspan=2, padx=(0, 0), pady=(30, 30))
+
+        ngram_check_label.grid(sticky="N", column=0, row=1, padx=(20, 0), pady=(10, 10))
+
+        select_all_checkbutton.grid(sticky="W", column=0, row=2, padx=(40, 0), pady=(0, 10))
+        n_to_ngram_checkbox[2].grid(sticky="W", column=0, row=3, padx=(40, 0), pady=(0, 0))
+        n_to_ngram_checkbox[3].grid(sticky="W", column=0, row=4, padx=(40, 0), pady=(0, 0))
+        n_to_ngram_checkbox[4].grid(sticky="W", column=0, row=5, padx=(40, 0), pady=(0, 0))
+        n_to_ngram_checkbox[5].grid(sticky="W", column=0, row=6, padx=(40, 0), pady=(0, 0))
+        n_to_ngram_checkbox[6].grid(sticky="W", column=0, row=7, padx=(40, 0), pady=(0, 0))
+
+        data_size_label.grid(sticky="W", column=0, row=8, padx=(20, 0), pady=(30, 0))
+        ngram_data_size.grid(sticky="W", column=0, row=9, padx=(20, 0), pady=(5, 0))
+
+        target_language_label.grid(sticky="W", column=0, row=10, padx=(20, 20), pady=(30, 0))
+        target_language.grid(sticky="W", column=0, row=11, padx=(20, 20), pady=(0, 0))
+
+        filepath_label.grid(sticky="W", column=0, row=12, padx=(20, 0), pady=(30, 10))
+        show_filepath.grid(sticky="W", column=0, row=13, padx=(20, 0), pady=(0, 10))
+        get_savepath_button.grid(sticky="W", column=1, row=13, padx=(0, 20), pady=(0, 10))
+
+        generate_dictionary.grid(column=0, row=14, columnspan=2, padx=(0, 0), pady=(30, 30))
+
+        about_help_buttonframe.grid(sticky="E", column=0, row=15, columnspan=2, padx=(20, 20), pady=(0, 10))
+        about_button.pack(side="left")
+        help_button.pack(side="right")
+
+        # Black turn on formatting
+        # fmt: on
+
+    def get_save_file_path(self):
+        self.savepath = filedialog.askdirectory()
+        if self.savepath:
+            self.filepath.set(self.savepath)
+
+    def generate_ntrans_dictionary(self):
+        self.user_choices = {
+            "save_path": self.filepath.get(),
+            "included_ngrams": self.user_chosen_ngrams,
```

Comment:
Keeping `user_chosen_ngrams` variable up to date is fragile: you must make sure that you get every `append` and `remove` right. This is a typical problem: you have the same data in two places (the `BoolenVar`s and the `self.user_chosen_ngrams` list), and keeping them in sync is difficult.

Consider just taking the `n`s where the corresponding value is set to true:

```suggestion
            "included_ngrams": [n for n, var in self.checkbox_vars.items() if var.get()],
```

Now you can delete `self.user_chosen_ngrams`, because it is no longer needed for anything.

Reply:
Genius!

For this, I need to work with self.select_all_ngrams() and self.update_ngram_checkbox() as well